File: Python/Problem4.py
class Problem4:
    def shortestPalindrome(self, s: str) -> str:
        if s.isspace() or s == "" or len(s) == 1:
            return s
        s_mirror: str = s[::-1]

        step: int = 0

        while True:
            test: str = s
            temp = s_mirror[:step]
            test = temp + test

            is_odd: bool = len(test) % 2 == 1
            plus: int = 1 if is_odd else 0

            if test[: len(test) // 2 + plus][::-1] == test[len(test) // 2 :]:
                return test

            step += 1


test: str = "abbacd"
expected: str = "dcabbacd"

p4: Problem4 = Problem4()


print(p4.shortestPalindrome(test))

File: TypeScript/Problem4.ts
function shortestPalindrome(s: string): string {
    if(s.trim().length === 0 || s == '' || s.length == 1) return s
    let s_mirror: string = s.split('').reverse().join('');

    let step: number = 0

    while(true){
        let test:  string = s
        let temp: string = s_mirror.slice(0, step)
        test = [temp,test].join('')

        let is_odd: boolean = test.length % 2 == 1
        let plus: number = is_odd ? 1 : 0

        let leftHalf: string = test.slice(0, Math.floor(test.length / 2) + plus).split('').reverse().join('')
        let rightHalf: string = test.slice(Math.floor(test.length / 2), test.length)
        
        if(leftHalf == rightHalf){
            return test
        }
        
        step += 1
    }
};